### ISSUE TITLE
Xposed presence

### ISSUE THREAD

**Issue by pylerSM** [pylerSM/NoDeviceCheck#3]
Need to know whether Xposed has any impact on "success/fail" state. If yes, we can hide it.

**Comment by hkpD** [comment]
It seems that Xposed does not have an impact on the "success/fail" state because I tried completely resetting my phone and tried adding a card in Android Pay and it still would not let me because I was running CM12.1 without root and Xposed

**Comment by pylerSM** [comment]
Yep, right I think

**Comment by Entropy512** [comment]
@haloali That's not really a good test of this, since SNet is pretty much an "if anything fails, it's a total failure" - so flipping Xposed on/off in a configuration that fails anyway tells you nothing.

The key would be to install Xposed on a firmware that normally passes SafetyNet checks and see what happens.  I have guests visiting at the moment, I'll try this on my Z3 when I get more free time.  (I'm running rooted stock for now, haven't had time since vacation to return to Omni.)

**Comment by Oggie7797** [comment]
I concur. Checking for xposed on a system that's guaranteed to fail is a poor test.

**Comment by sovanyio** [comment]
Wondering if the xposed check is as simple as scanning the logcat for the xposed string.. could be as simple as turning off logging for xposed.

**Comment by Entropy512** [comment]
@sovanyio What xposed check?  Do we actually have confirmation that there is one?  The original issue is just to determine if it does - I suspect it doesn't.  There's no mention of xposed in the SafetyNet teardown at https://koz.io/inside-safetynet/

**Comment by sovanyio** [comment]
You're right, its not explicitly called out. But it seems like another huge target that Google would check for in addition to su. I'll try installing xposed on my gpe rom later to see if that alone trips it.

**Comment by bengalih** [comment]
I can confirm that Xposed framework blocks Android Pay from working.
On my HTC M8 (bootloader unlocked/rooted) I could not get android pay to work by disabling root, installing RootCloak and No Device Check.
Only AFTER fully removing Xposed (and keeping root disabled) could it work.
So the presence of Xposed is certainly influencing Andoid Pay.

Can this be made to block it?

**Comment by sovanyio** [comment]
Second. I smell a coming cat-and-mouse game with this though. If they are just doing a simple logcat check for 'Xposed' and we get around that, they could certainly detect the filesystem modifications and then what?

**Comment by Entropy512** [comment]
Thing is it may not explicitly be looking for Xposed - it may simply be looking for modifications to the app_process binary - Someone really needs to look into how an attestation is generated, I'll try to do a bit of looking tonight.

**Comment by Entropy512** [comment]
OK, so I've traced through some of the SNet code at https://github.com/pylerSM/NoDeviceCheck/issues/2 and have hit a brick wall...

It appears that the attestation info is part of the GBundle whose source is at com/google/android/snet/GBundle.java - see snet_attest_bundle - This appears to be somehow derived from GBundle's paramBundle

com/google/android/snet/Snet.java indicates that this GBundle's paramBundle is the paramBundle passed to the Snet class when it is created.  It appears that this is passed by enterSnet or enterSnetIdle.  However, I'm not seeing what is calling enterSnet - I'm guessing that this is probably somewhere within Google Play Services.

**Comment by humulos** [comment]
I don't know if my input is helpful, certainly don't understand the mechanics behind any of this but I can definitely confirm that simply having xposed installed does prevent Android Pay from working, so something that it affects is being checked. Here is what I checked, all with the Bootloader locked via BootUnlocker
1. Xposed enabled, all modules on (including Rootcloak and No Device Check), SuperSu enabled: Android Pay Fail
2. Xposed enabled, all modules on, SuperSu disabled: Android Pay Fail
3. Xposed enabled, all modules off, SuperSu disabled: Android Pay Fail
4. Xposed disabled, all modules off, SuperSu disabled: Android Pay Fail
5. Xposed uninstalled, all modules off, SuperSu disabled: Android Pay Pass

Thank you for your work on this, I really miss using NFC payments, but Xposed is far more valuable to me.

**Comment by pylerSM** [comment]
Try to just uninstall Xposed Installer (do not uninstall Xposed binary) and tell us result.

**Comment by rich0** [comment]
I suspect the key is going to be in messing with how the attestation is performed.  It doesn't sound like they're using signature chaining (Palladium style) and remote attestation, so it is probably defeatable.

I've seen statements online (which may or may not be true) that it is checking for modifications to the system partition.  I'd be curious as to whether just sticking a random text file in /system or installing busybox/etc without rooting or installing xposed causes failures.  I'd think that doing a full hash of system is going to be too slow, so they're probably looking for specific markers which we could evade.  

The other approach is just patching the software doing the attestation to just always pass, or intercept the call to it.  Or just patch Android Pay to not do the check.

As long as they're not doing full remote attestation with signature chains we should be fine.  If they go that route then you're stuck cracking TPMs.

**Comment by Entropy512** [comment]
I'm fairly certain some of the Android Pay checks are server-side - see the link I posted 20 days ago in this thread.  That explains why patching the response isn't working.

The attestation code does not seem to be in the snet bundle, so it's probably somewhere in GMSCore itself - I haven't had time to look for it.

Extra files may not trigger it, but I would NOT be remotely suprised at all of a critical core component such as the ART engine were being, at the least, checked for size and modification time.  (I am assuming "xposed disabled" above involves a modified engine still being present but not "doing its thing" but I could be wrong, I'm a bit rusty on my Xposed.)

**Comment by humulos** [comment]
As requested, I tried just uninstalling the Xposed Installer, Android Pay wouldn't add my card before or after rebooting. SuperSu was disabled and Bootloader was locked for this attempt.

To answer Entropy512 on xposed disable, you can see this thread on XDA for details of the zip I flashed, which as far as I understand it was just adding an extra file to \data\data\de.robv.android.xposed.installer\ that stops xposed from properly executing. 

http://forum.xda-developers.com/xposed/xposed-framework-disabler-incase-t3100669

**Comment by Entropy512** [comment]
OK, in that case, you still have a modified system file.  I'm 90% certain that it's a change to that file that's triggering the failure.

My Xposed-fu is extremely rusty, but an interesting trick might be:  Determine the file size and modification/creation time of the files Xposed modifies from a stock firmware.  (Challenge here:  This will be different for every device and every build).  Then use xposed to fake these out if anything checks them.  Won't work if something hashes the file contents though.

It won't work across all devices but would give hints as to what does/does not trigger failures.

I suspect that a long-term solution will have to be to hook whatever generates the final attestation data result, and replace this with something that matches the device running stock firmware.  A database will be needed for each device/build to enter this result...  :(

**Comment by humulos** [comment]
To preface this, I again have no real knowledge about how this works, so my ideas may have no basis or plausibility, but...

When xposed does its thing, does it create backups of the system files or does it just directly edit them without copying the files first? If it does create backups, could the module find way to have apps look for the backup files instead of the current system files?

No clue if that is even remotely logical, but just thought I'd throw it out there.

**Comment by rich0** [comment]
No argument that there is a server-side component.  It sounds like something local is calculating a hash (what it is based on is unknown, but it obviously picks up su and xposed), and sending it to the server for validation.  

However, as long as there isn't a TPM-backed signature chain involved it shouldn't be straightforward to defeat once it is worked out.  You could intercept the framework/system calls from the checking program and feed it unmodified data, or you could intercept the calls to the validation routine and just feed back a valid hash to the caller.  

If a TPM were involved you'd be hosed, since it would hash the kernel and sign that, and the kernel would in turn either only run controlled binaries or hash whatever it does run and sign that.  You could never get a complete chain of trust without running stock everything from the bootloader to the executables.  

I haven't heard any discussion of kernel-related issues, so it sounds like the kernel itself is not being checked (which definitely would rule out TPM checks).  That is an obvious weakness right there, if indeed you can get android pay to work with a custom kernel.

**Comment by gitchrodd** [comment]
So i flashed factory image MRA58K on my Nexus6. Then Elite kernel. with unlocked bootloader no root or recovery.  I can get android pay to add a card and i tested at work was able to make a purchase. So at least one custom kernel had no effect. And my unlocked bootloader also was not checked?

**Comment by Nemisor** [comment]
All I have to add is: keep trying please! We shouldn't be denied Pay just because we're running modified phones.

**Comment by d8ahazard** [comment]
Hi guys!  I just wanted to weigh in on what rich0 was talking about - 

I do a custom GPE ROM with Android Pay.  We've also discovered the same Android Pay will only work with SuperSU disabled and Xposed uninstalled.  What is of note - is that, yes, Pay _will_ work with a custom kernel.  My ROM allows beyondstock and elite kernels, and provided xposed is gone and SU disabled, it works fine.  

Additionally, I don't think it's based on the system itself being modified, as I've got a number of custom mods to my /system partition, and again, pay works just fine.  

I've been meaning to try playing with the code for this myself as I'm now doing Xposed development as well...maybe tonight I can do some more experimentation with it.  My guess is that something somewhere is either looking for XposedBridge.jar in framework, or perhaps the symlinked binaries in /system/bin.  

I've also decompiled the pay app and had a look with no success, but haven't taken a good crack at GMSCore.  Given what I'm reading in here, I have a few new ideas I will try.  If I come up with anything, I'll def. post back here.

**Comment by d8ahazard** [comment]
Update:  So, I forked the module, added some logging to see what other checks may be happening.  Found some interesting stuff.  Log output is here:

http://pastebin.com/iAPi5piR

OH SNAP!  There exists an apk in /data/data/com.google.android.gms/app_dg_cache/BIGSTRING/t witha  package name of com.google.ccc.abuse.droidguard.  Betcha this is the culprit!!

Edit:  I've added more hooks to log every string and file android or wallet is checking, plus the start of hooks to look at what DroidGuard is doing.

**Comment by pylerSM** [comment]
Good work :)

**Comment by djmigues** [comment]
I have Xposed enabled on my phone (Verizon HTC One M8) and I am able to add cards and make purchases as long as I disable root in SuperSU app.

**Comment by Devo7v** [comment]
Were you able to figure out what DroidGuard is doing?  Hopefully this is the check that needs to be bypassed.

**Comment by th3fallen** [comment]
I'm not able to add cards in pay with only xposed installed. went so far as to stock my device and start from a clean slate. Using SafetyNet Helper (app) i get this response with only xposed installed. (unrooted)

SafetyNet request: Success
Response Validation: Fail

Error Msg:
Response payload validation failed

So it appears your module is missing something somewhere....

**Comment by bgarlock** [comment]
As soon as I uninstall Xposed, Android Pay works (even with systemless root). I'm stock Moto X 2015 PE.  It seems that anything installed to /system will prevent Android Pay from working.

**Comment by d8ahazard** [comment]
So, I have verified that some of the "Xposed Presence" stuff is simply good-old file checks.  

https://github.com/pylerSM/NoDeviceCheck/pull/5

Using my modified version, I've now got it logging the checks it's doing for Xposed, and theoretically, bypassing them.  I don't have a stock/official system image installed right now to test further, but this should definitely help.  

I'd advise merging the pull request, and building on the methods I've added...

**Comment by humulos** [comment]
If you can get me an APK, I can test this tonight for sure, and would definitely love too. Stock Nexus 6P, February Security Update, Systemless Root with Xposed.

**Comment by d8ahazard** [comment]
You should be able to download the zip from my fork, extract, and run "gradlew build all" in the root of that folder to compile - I made sure to include all gradle files for that specific reason.  

If that doesn't work for you, lmk, and I'll upload an apk.

**Comment by humulos** [comment]
I've never done such a thing before, but rather than take the easy way out, I think I would like to try :-)

**Comment by Nemisor** [comment]
I'd like the easy way out, if that's possible? Stock S6 here, December update, but I could patch to February if needed...

**Comment by humulos** [comment]
Okay, so building the Apk was actually super easy (once I fixed my environment variables) so that's nice! Tried installing it (there were 3 apks, I chose the one called app-debug, hopefully that was correct), and enabled it in xposed. Rebooted. Strangely, the Xposed app was no longer in my apl drawer. Attempted to install it, got an error that I was out of space. I have over 30 gigs free, and larger apps from Play Store install just fine. Ignoring this, Gravity box was clearly still working, so I thought maybe this apl would still function too. Tried adding a card, but it failed. Disable root, still failed. Will re-root and reinstall Xposed when I get home and try a few more variables.

**Comment by d8ahazard** [comment]
So, the very latest commit I have in there causes the weirdness with the SDCard issue.  Just noticed that - I'll fix it momentarily.  If you build from the version Pyler just merged into his, it shouldn't be as wonky.

However, Xposed missing is actually a result of the module doing it's job.  It's there, and it works, but all calls to see if it exists are currently blocked - as intended.

I really want to point out that I don't expect my new additions to make Pay work - yet.  The intent is to get a better look under the hood of SecureNet and see what it's looking for, so we can stop it.  For example, if you now take a logcat while trying to authorize pay - you can see where the app is checking for xposed, su, etc.  

So, that's still _progress_.  It just doesn't work yet.  

I'll be on this all night, probably.  Will post back as I have more changes in the pipe.

**Comment by humulos** [comment]
Okay, gotcha, that does make sense. So in that case, what else can I test that would be of assistance?

**Comment by al3xtjames** [comment]
@d8ahazard: Great work with your commits! The logs are very useful, it's easy to see what SafetyNet is checking. [Here](https://ghostbin.com/paste/4mbb9) is some of the logcat output from a Nexus 6P running CM13 (with built-in root enabled, which IIRC is present on the system partition) when attempting to add a card in Android Pay.

**Comment by d8ahazard** [comment]
Hey, thanks!  Looking at your log, I noticed something that confirms a suspicion I was working on last night.  

If you take a peek - every time it checks for /su in /system/bin or /system/xbin, it's doing a .listFiles for the containing directory.  I betcha that the result of this check is what triggers the check for /su itself, and what fails the resulting checks.  

So, while you can see the result of me replacing the boolean checks (exists, isfile, etc) with a "false" result, I have not yet implemented a catchall sort of method that will work for all of the various directory enumerations we're going to have to anticipate.  

I want to have one method that will have to loop through the resulting array, check for any of the "keyword" files, and then remove them from the array, return the modded array so that it never even knows the files were there.  

I bet you that if we get this, there's a chance we'll beat SafetyNet.

**Comment by MistarMuffin** [comment]
I have a Galaxy S6 running systemless root via SuperSU 2.67 BETA. After setting CHMOD 751 on /system/bin I can use Pay. I installed xposed and of course it breaks SafetyNet.

I reviewed the logs that theracermaster posted and saw that in addition to SU related files, Google Play Services appears to be checking the following files on /system:
app_process32_xposed
app_process32_original
app_process64_xposed
app_process64_original
xposed.prop

I installed the RootCloak module as it has a keyword blacklist which is supposed to block specified apps from even checking if files matching the keywords exist. I added the files above to RootCloak and specified Google Play Services as a target but it made no effect.

Is there any reason it should be more complicated than this?

As a temporary workaround could a custom version of xposed be compiled that simply uses different file names that would pass the specific isFile checks? Granted this would not be a permanent solution as it would just create a cat and mouse game.

**Comment by d8ahazard** [comment]
I've not had a chance yet to work on a method to fudge the check - but I bet if you look closer to your log, you will see that right before the check for the specific files, there is a list() or listfiles() command executed against the containing directory for the files in question.  

My suspicion is that we need to hook these methods as well, capture the array of files being returned, run a for loop against every array item with our keywords, and if the item does _NOT_ match the keyword, drop that in a new array.  

Essentially, we want to create a scrubbed array of all the files in /bin /system/framework etc. that are NOT our checked files, then return that.  

If my suspicion is correct and we write a method that does this correctly, I bet you won't even see any more checks for the individual files, as the list command won't trigger the next checks, which...by the time they run, it's too late.

**Comment by pylerSM** [comment]
@MistarMuffin can you share log?

**Comment by MistarMuffin** [comment]
Sorry, I took no logs of my own. I just reviewed theracermaster's logs for which files Google Play Services was looking for and attempted to block them with RootCloak. That being said, I think d8ahazard is right and that RootCloak isn't working because GPS initially asks for a list of all files in /system/bin and doesn't make queries for specific files until it already found them in the file list. The jig is up before RootCloak even has anything to block.

Is there anyone experienced with writing Xposed modules that can do what d8ahazard suggested? Intercept the file list array that GPS requests and remove the suspect files and return the modified array? Unfortunately that is not me.

**Comment by gamer765** [comment]
I forked this module and added the appropriate loops for checking the array for the keywords and if found, omit them from the result. I've modified the result of list() and listFiles() to not include them, which works because now in the logcat I do not get instances of
`Found matching string - /system/bin/su. Caller: com.google.android.gms. Checktype: getPath
NoDeviceCheck: Found matching string - /system/bin/su. Caller: com.google.android.gms. Checktype: getPath
Method hook executed (getPath) for file /system/bin/su, post-result=
Found matching string - /system/bin/su. Caller: com.google.android.gms. Checktype: exists
NoDeviceCheck: Found matching string - /system/bin/su. Caller: com.google.android.gms. Checktype: exists
Method hook executed (exists) for file /system/bin/su, post-result=false`
in my logcat.
The only app that still calls /system/bin/su is the MyTMobile app, but thats probably because it might be hardcoded to search for that specific file in that specific directory. I'll post my fork shortly. Also i've noticed that removing the keyword files from the result of list() and listAll() hides the files from the Android OS itself; File managers will still account for the correct amount of files in the folder but our keyword files will have blank names and we cannot view/open them.
Also should note that Android Pay still does not work.

**Comment by Nemisor** [comment]
Thank you for this, gamer765, hopefully one step closer to getting Android Pay eventually working =)

**Comment by d8ahazard** [comment]
It may also help to note that I just read something about safety net, where
it may also report on either the Verity state of the device if it has those
checks, or whether or not bootloader is unlocked.  Also, I would expect it
to look to see if selinux is set to enforcing. So we probably want to
figure out what methods they do to check for those.

**Comment by gamer765** [comment]
my fork is posted with my changes. It's not pretty but it works; original code had more for loops cycling through the old array and new array logging each index value, our special files were no longer listed.
Also forgot to mention that since our files of interested are no longer listed, you may lose root access, as Android doesn't know the files exist and therefore can't access them.

**Comment by gitchrodd** [comment]
Safety net will not fail for unlocked bootloader or selinux policy. Fyi

**Comment by gamer765** [comment]
Here's a little insight to why the Xposed module doesn't work, we're patching out java calls in the java side of Google Play Services, but Google Play Services downloads a few extra files, snet.dex, snet.jar, and metadata. The file of interest is snet.dex which is not an Android DEX file, but rather an executable ELF file. Also it loads XposedBridge.jar into its classpath. I believe that by patching those native checks out we beat SafetyNet
![xposed](https://cloud.githubusercontent.com/assets/473362/14235547/913f822e-f9cf-11e5-823e-ab92a6964918.png)
![sucheck](https://cloud.githubusercontent.com/assets/473362/14235548/91452b84-f9cf-11e5-866a-d8a63c02d1cf.png)

**Comment by chenxiaolong** [comment]
@gamer765 Are you sure that's not an optimized `oat` file created by dex2oat? Those are stored as ELF binaries as well.

**Comment by gamer765** [comment]
well you might be right but this file is most certainly not an oat file created from snet.jar It's been a while since I messed with ARM disassembly stuff.
Actually I just double checked, yes it is an oat file created from snet.jar. I'm low on sleep.
![sucheck](https://cloud.githubusercontent.com/assets/473362/14235601/44027046-f9d1-11e5-858a-980419ebb952.png)

**Comment by chenxiaolong** [comment]
@gamer765 On my device, `snet.jar` exactly matches `snet.dex`. I disassembled the oat file with https://github.com/testwhat/SmaliEx as well as the snet.jar file and they result in the exact same smali code.

**Comment by gamer765** [comment]
yeah and play services is calling snet.dex by reflection

**Comment by d8ahazard** [comment]
This is awesome! Currently, I'm on vacation in Florida for the next week.
But when I get home, I definitely want to chat with you on Hangouts, I
think maybe if they you if you and I put our heads together, we can crack
this thing quite quickly.

**Comment by Devo7v** [comment]
@d8ahazard , @gamer765 : Any luck?

**Comment by Maxr1998** [comment]
I just had a totally different idea: instead of patching around the snet.jar and .dex until we hide everything they try to find, wouldn't it just be possible to replace the dex with our own version which doesn't check at all and passes all requests?

**Comment by Entropy512** [comment]
No, because it's possible for the device to send attestation data to a remote server to be verified (there was a good analysis of SafetyNet that found this aspect of it).  That attestation data isn't a "pass/fail" - it's information the server uses to determine a pass/fail scenario.

See Android Pay.  Patching local pass/fail tests doesn't work (proven by early attempts at NoDeviceCheck), because I'm 90% sure Pay sends attestation data to the server - which fails its checks on that data.  So you can't just patch a yes/no result, you need to fake information such that an attestation looks "good" to a remote server.

**Comment by Maxr1998** [comment]
But would it be possible to capture the tcp packages of a device that passed SafetyNet, and then use that data to replace parts of the sent data?

**Comment by marzika** [comment]
The whole data sent is hashed and signed you cannot extract parts of it. Secondly, we (devs of Snapprefs) were able to replace a failing attestation with a previously accepted one, however the whole attestation expired after a while. It was back in November, haven't looked into it since then.

**Comment by Entropy512** [comment]
Interesting...  Do you happen to have any examples of what you did there?

It would be interesting to see just what data is in the attestation, since that would hint at SNet is looking for to build that dataset.

**Comment by nayanbhana** [comment]
@entropy512 I think I have some of it. I will look for it today

**Comment by Entropy512** [comment]
As to the attestation expiring, from https://developers.google.com/android/reference/com/google/android/gms/safetynet/SafetyNetApi#public-methods

"When you request a compatibility check, you must provide a nonce, which is a random token generated in a cryptographically secure manner. You can obtain a nonce by generating one within your app each time you make a compatibility check request. As a more secure option, you can obtain a nonce from your own server, using a secure connection."

A server-verified attestation will likely expire after one use if the service is designed correctly - it will use a different nonce with every attestation.

**Comment by Entropy512** [comment]
BTW, a potential method for hiding Xposed, based on the positive results someone had with "systemless xposed"

If I recall correctly (It's been a while, I'm rusty), Xposed keeps a backup of the "original" app_process in /system/bin - that backup's presence is likely one of the triggers of a failure.

The second thing is that SafetyNet likely checks aspects of the app_process binary (size, modification time, etc.) - Hooking anything that tries to check the file properties of app_processed and making the file appear to be the same size/modification time as stock app_process might cloak Xposed from Android Pay.

**Comment by Entropy512** [comment]
@gamer765 The fact that file managers are reporting the correct number of files and you're seeing the "cloaked" files in a file manager with blank names is an indication you haven't fully cloaked the files.

If I'm reading your code correctly, I think that arraySize will always be the number of files in the original result - you are going to have blank strings in the result.  I'm about to make a few comments on your github commit.

**Comment by nayanbhana** [comment]
@Entropy512 So it has been a while since I looked at this but I think the hash @marzika is referring to is Protocol Buffers. https://developers.google.com/protocol-buffers/ 

For Snapchat we used to generate our own protobuf and the server would return an attestation, the format I used there can be seen here. [https://i.gyazo.com/f4fee815c2ca5b5c70940f135a983e79.png](https://i.gyazo.com/f4fee815c2ca5b5c70940f135a983e79.png)

The attestation we generated would always return a fail but at the time Snapchat wouldn't block your login. I think it was about November when Snapchat started blocking the login and I haven't looked at it since.

**Comment by Entropy512** [comment]
Hmm.  Perhaps Snapchat was initially ignoring the nonce when you did the replay attack?

Seems like not much information there, unless unknownData is a multi-byte affair?

Do attestations which usually "pass" contain any suexec submessages, or does presence of one of the suexec submessages (the "repeated suexec" line in dataContainer) always cause a failure in your experience?

Did you notice any patterns to the unknownData bools?

What was the typical value of the droidGuard string?

Edit:  I was really expecting the attestation to include a hell of a lot more information than that...

**Comment by nayanbhana** [comment]
droidGuard string 
`Cgbp9kfDzmHYAQE4YRoWCJDI_r3______wEQuJTS-vr_____AaoBFgi8u_-0_f____8BEJ7mkq_6_____wHoAa-Zm7P4_____wHoAbqB_d4FOGs4SxL8CZJhg1mxAi2L1xGeKr-lp028We0xQzTGEBz3tNVEZYMaTZ1APVFk6iiRbcj45WlYaYu38C7M_dnU3BnLA8-IRFucCGJg7zjK4CxRA8bIA2xILFfa8Nt0493ZazjpKS_Onqs6W9nrx9DxG3SdcErWiYuYbFq_VRgLF-VLUBcvr7LlzJZ4ah8Qqo3i9LrBI5MUigZkrYsZgJTE5K5Sst11m0xITkeg8R77IJFotTiQxPQopD2Fyxja96Die5DosiW_2uVCkJ5ms2dpLi6F6QTmwmL62WW_wS1SXE836B1bTXSKeNu3yyJDAK3OOsLpriMMHRvqdeCWCJ4A-oG5JPC0U8-OlrizRB_iqcpVv-x4GbS2Y3gOUZJrKl56mKYxe3YxMVMjKDAHWTIoyEIXB_vH6N5idxmpXn82IJbCtERVxwuX7WEQMewxsGaWwLyqrbMGXmoIMNzrdQULyd-BU-mkQV8uzSOO3wuer24eCFGVDuCBqvhMkeCeezwcn0t_KyQw5r2JGRvZJ49abqs38cI7g6UI-kVFQsho8DL1k0nITBM_DqfxJVUUBbuek_LwRMsy5JJiudy3J0n7xSW351mnP0WTb9eeshxKgjwT2wxE9xNSRF92LdaV7ZsGHhI5PP83erN9niQ1aBvlEGzU4dGVlMG10p6UXpJj2V88-0Wk3r_Hk3ZCLgtj1Q0BO1Wrjpc1WerqaH2kSXFPH7WGFMtfBcpovmFm1JMho-LiwNiKUPFi8Tkj69Gr3pp0CUrCdIb_HUkWDEJdLFJ8PgjRS7twnAN8f8I7qufxj2aEEQfDxsKL5l9GxoD4bM6Q6iP-gzmlvILLA0sJgCQr1u2ytjAnJnBLBxxGo8KgGIRKMs1NJnZwAc71Ur9TCo3KzNPCe-g98Y6noF_CxZhFWgJDnfHyMSDUeZv1vr6h_ogxJEaRPEn-4OZlxekJ5rUZV8JS_G-IpA7QXWv2KZw20Y4uS_qofqv7ogH-xa7UZXbRoY_H3nAA_q4id_HoUMUepJ0KICOx6Xc_R0tV_kjzxQ1ECiFWgUjrxieU6Z3fSES2wFGcHaRUdWKnS1PxprMq_IEbSQrbvAZvh5LS5HyIcAlGeOfv1GuQt2wtk-DZWgDdr-TnZzZBsAwhfu82BI1Y2P8RrtuV0IdEjlj4p382oluC7o3UOd10CiXVdxocjff7kC_TDEfxcWJGQFnQih5zHoxwUeyjaiS1bMwCZtlzaX7xyfgDTx-QtoxeViHG8CSKhs5Y2ZZYaLK1Q6eC_RbUMwjl0_vGJnTFgDO4Zz42GiWh7oyzYNtHGXp3jTYy9w5YQg6U0_w65TthAH4mXACvoqiQEhCt_uI5jqbhg0pCktxbfc78RfCS6BPD9XJL6f54ig95PnVR-3PSEqihMdJ_rVaK-l5-wr8QdHu5prgWDHlSPSUNXnBC-BuBCgBrXA6BjU8KmHP3bQujbrycVU5Ko1kzeLg46_pbOZ2EecelnyUMfMpPoHx7ax9Sfp7d2-T_WjPcfCB10ZIwFB8102To3UNFDPrwpxH9fslWb6PuY1VuAQ1VW79K-xgGqdJoi418hycjgOqvqhZf_kn_QxmOzRfJbFkZW8znaXDZq7RuaBZWOkVARgy-xwh3T0v86C1QdMfCTVb0ceOMe3KElqUlLgiYfSV9gOMOrMIuZXj3idSJkQAzRvmeU-JyCjFkNvFLG0Q'`

**Comment by burinov** [comment]
FYI: SafetyNet calculated hash on /system, so installing Xposed in /system will fail.

There is Systemless Xposed, but snet detects that it is there, but... According to the some people it is possible to fix too...

Please, tell me how you get snet sources? snet.jar? snet.dex?

I also want to join this party and be useful, but I don't know how to approach snet yet.

**Comment by hahaopsmeow** [comment]
According to "Android Security 2015 Year in Review" (google for it, it will be top PDF link). 

Page 20:

> It (talking about snet) also provides a way to measure the diversity of the Android ecosystem: SIC 
> has identified ed over 175,000 unique system partitions (and over 60,000 system partitions that 
> have more than 1,000 active devices).

It means that SafetyNet calculates hash on /system, so installing Xposed in /system will fail.

Here is more info: https://koz.io/inside-safetynet/

---

The only way I see is to use Systemless Xposed. But there is another problem: SafetyNet sees Xposed and it is immediate failure. It is possible from one of the following methods: http://d3adend.org/blog/?p=589 (and few more as I can count).

But, according to the some people it is possible to fix too...

Unfortunately I am very new in this topic and need some of your help to proceed:

Please, tell me:
- Where and how you obtain snet sources?

**Comment by rich0** [comment]
Seems like the two obvious approaches are the change SafetyNet so that it reports stored hashes instead of actual ones, or to change the kernel so that when SafetyNet reads /system it gets a copy of the original /system.  If it is actually hashing it at the device level (and not just individual files), that would probably require storing a copy of all blocks that are modified.  If it is just hashing all the files then storing backups would be fine.

This is basically just the usual rootkit / stealth virus approach.  The only way Google could really bypass that is by using remote attestation backed by a TPM (which I believe most of the ARM chips actually support these days).  

If we're going to muck around in the kernel this much a systemless approach probably makes more sense, and then we just have to store a copy of an unmodified kernel someplace and if something tries to read the kernel partition just point it at the original.

**Comment by hahaopsmeow** [comment]
More input for Systemless approach:

> SafetyNet will detect the current running app_process (zygote), so even if Xposed is 
> installed systemless, it can still detect you're using a heavily modified binary. 

Any ideas?

---

It seems that: systemless xposed with "detection" fix and we are good to go!

**Comment by rich0** [comment]
> > SafetyNet will detect the current running app_process (zygote), so even if Xposed is 
> > installed systemless, it can still detect you're using a heavily modified binary.
> 
> Any ideas? 

Seems like just another reason to patch the kernel.  SafetyNet only knows what the kernel does it.

Again, Google can escalate that one further by planting their checks in the firmware with TPM support.  Ultimately whoever controls the hardware owns the device.

**Comment by marzika** [comment]
Another way to detect the framework without detecting the files directly: https://github.com/rovo89/XposedInstaller/issues/93

Also, you might have noticed that when using Xposed Android Studio throws up bunch of "Warning: linker: app_process has text relocations. This is wasting memory and is a security risk." which should be theoretically solved by using -fPIC when compiling. It can also be used to detect the modified app_process.

**Comment by Entropy512** [comment]
@nayanbhana Thanks.  I think the droidGuard string is key here.  Need to figure out how that's being built - that name is something that can't be obfuscated so that's probably a good starting point in looking through decompiled source

@hahaopsmeow There's no way it's hashing all of /system - that would take too long.  It appears to be collecting file lists, and doing a "deeper" inspection of some sort on select files.

**Comment by Entropy512** [comment]
Systemless xposed uses bind-mounting - so when it's "active", /system/bin/app_process is the "modified" one.

**Comment by hahaopsmeow** [comment]
@Entropy512 you are right. Whole system hashing makes no sense.

> **system_partition_files**
> Collects various information about a configurable list of files of interest under /system. Information 
> includes all permissions, names of files, symlink targets, selinux security context, hashes etc. 
> Currently the list of files of interest seems empty. 
> It is interesting that the framework makes use of red herring in order to confuse attackers: along with 
> the files of interest, a configurable number of random files can be accessed in the same way. 
>  Great stuff, red herring approaches are commonly used in advanced app protection products. 
> However this is not enabled by default.

More info that proves it:

> How is the /system hash created?
> SafetyNet runs an process that recursively walks "/system" and calculates a HashTree over its contents. 
> For every file it encounters it captures meta-information (timestamps, permissions, selinux context etc) 
> and its SHA256 hash into a local data store. For every directory, it generates a hash that considers the 
> store entry of every file inside the directory. If there are hash mismatches between previous and current 
> recursive walks over /system, the offending files are entered in separate lists to be audited.
> 
> The LOG SYSTEM PARTITION FILES module continues to include the results of the SystemPartitionFileFinder sub-module. 
> As a reminder, this module retrieves the status of various files in /system. The list of "files of interest" 
> is configured over the air. Currently, the following files are checked, along with 5 random files:
> 
> ```
>   /system/app/providerdown.apk,
>   /system/priv-app/cameraupdate.apk, 
>   /system/app/cameraupdate.apk,
>   /system/priv-app/ThemeManags.apk,
>   /system/app/HTMLViewer.apk,
>   /system/app/com.android.hardware.ext0.apk,
>   /system/app/com.android.wp.net.log.apk,
>   /system/app/com.google.fk.json.slo.apk,
>   /system/app/com.google.model.mi.apk,
>   /system/app/SettingProvider.apk,
>   /system/app/SecurityCertificate.apk,
>   /system/app/LiveWallpaper.apk,
>   /system/app/BatteryControl.apk,
>   /system/app/Models.apk,
>   /system/bin/.daemon,
>   /system/bin/.daemon/mis,
>   /system/bin/.daemon/nis,
>   /system/bin/daemonnis,
>   /system/bin/nis,
>   /system/bin/.sr/nis,
>   /system/bin/.sr,
>   /system/bin/.memnut,
>   /system/bin/.suv,
>   /system/bin/.sc/mis,
>   /system/bin/uis,
>   /system/usr/.suv,
>   /system/xbin/.memnut,
>   /system/xbin/.suv,
>   /system/xbin/ku.sud,
>   /system/xbin/.rt_daemon,
>   /system/xbin/.rt_bridge,
>   /system/xbin/.monkey.base,
>   /system/xbin/.ext.base,
>   /system/xbin/.like.base,
>   /system/xbin/.look.base,
>   /system/xbin/.must.base,
>   /system/xbin/.team.base,
>   /system/xbin/.type.base,
>   /system/xbin/.view.base,
>   /system/xbin/.word.base,
>   /system/xbin/.zip.base,
>   /system/xbin/.bat.base,
>   /system/xbin/com.android.wp.net.log,
>   /system/xbin/.b,
>   /system/xbin/.df,
>   /system/xbin/.c,
>   /system/xbin/.sys.apk,
>   /system/xbin/.ld.js,
>   /system/xbin/.ls
> ```

---

Some people reported that having systemless root and systemless xposed installed but disabled allows them to use Google Pay.

So, it seems that xposed detection in snet is the key.

---

Guys, where do you get snet sources?

**Comment by hahaopsmeow** [comment]
@Entropy512 you are right. Whole system hashing makes no sense.

> **system_partition_files**
> Collects various information about a configurable list of files of interest under /system. Information 
> includes all permissions, names of files, symlink targets, selinux security context, hashes etc. 
> Currently the list of files of interest seems empty. 
> It is interesting that the framework makes use of red herring in order to confuse attackers: along with 
> the files of interest, a configurable number of random files can be accessed in the same way. 
>  Great stuff, red herring approaches are commonly used in advanced app protection products. 
> However this is not enabled by default.

**Comment by hahaopsmeow** [comment]
@rich0 

> Seems like just another reason to patch the kernel. SafetyNet only knows what the kernel does it.

What exactly should be patched?

I think I we have sources of snet, it will be easier to understand work need to be done.

**Comment by Entropy512** [comment]
I think @gamer765 was on the right track - but at least unless he tried something he didn't push to his github repo, his method for cloaking files would've left evidence in the form of "blank" files at best.  With some changes, it might do the trick.

Except that in addition to hiding xposed's "backup" files, when SNet tries to probe app_process, that probe should somehow be redirected to the original app_process (e.g. SNet only sees the unmodified binary).

As to snet sources - decompiler.  I believe much of the attestation is assembled elsewhere though, last time I looked through them I couldn't find where it was being assembled - however that droidGuard string helps a lot.  The snet sources do reference a "droidguardResults" string...  hmmm

**Comment by Entropy512** [comment]
Actually, I think the DroidGuard service is somewhere buried in GMS, not in the snet package.

**Comment by Entropy512** [comment]
mergeFrom in com/google/security/cryptauth/lib/risk/StarguardData.java looks interesting

**Comment by pylerSM** [comment]
Which API  it uses to scan "/system/bin/app_processXX"? we could hook it, check for this path and redirect it to "/system/bin/app_processXX.orig" (original untouched file)

**Comment by hahaopsmeow** [comment]
> Which API it uses to scan "/system/bin/app_processXX"? we could hook it, check for this path and 
> redirect it to "/system/bin/app_processXX.orig" (original untouched file)

This won't work, because snet also detects where xposed is running or not. Please see below for details:

Please check it out (description of systemless xposed): http://forum.xda-developers.com/showpost.php?p=67074423

It says:

> **_Benefits**_
> **OTA:**
> - The system partition will not be modified or mount r/w, thus your Android devices are capable of applying OTA updates.
> 
> **Android Pay:**
> - Unfortunately, Android Pay won't work with Xposed enabled even if no system files are touched.
> - However, by simply toggle the disable toggle in Xposed Installer App, systemless method gives us the ability to re-gain Android Pay just in a reboot!!
> - Disabling system installed Xposed will not pass Android Pay check! Confirmed with my HTC 10 and Nexus 9.
> - Systemless Xposed is required for Xposed to live with Android Pay. Disabling won't trigger ART cache rebuild.

So, for me it seems that we need to hide from snet the fact that xposed is running.

I am about to test it now.

**Comment by Entropy512** [comment]
I'm fairly certain it's not detecting that xposed is running, it's detecting that app_process changed.

I suggest reading http://unix.stackexchange.com/questions/198590/what-is-a-bind-mount - specifically, "Unlike a hard link or symbolic link, a bind mount doesn't affect what is stored on the filesystem. It's a property of the live system."

This is how systemless xposed works - the modified (xposed) app_process is NEVER written to the system partition (which is why the claim that the system partition is not modified or mounted RW).

HOWEVER - when "enabled" - systemless xposed bind-mounts a modified app_process binary at the location of the "original" app_process binary.  So while the /system partition was never modified, _to any application it appears that this is the case_ when the bindmount is enabled (Xposed active).

@pylerSM I think that is exactly (along with cleaning up @gamer765 's directory listing code a bit) needs to happen.  Unfortunately I don't have time for a while to try hacking this myself, but I can read code and make some suggestions/ask questions based on what I see.  :)

BTW, if I'm reading things correctly, the DroidGuard string may be serialized protobuf data - see https://github.com/google/protobuf/blob/master/javanano/src/main/java/com/google/protobuf/nano/CodedInputByteBufferNano.java and mergeFrom in com/google/security/cryptauth/lib/risk/StarguardData.java within the snet sources (warning:  my snet sources are VERY old, generated by decompiling one of the first known snet packages)

**Comment by Entropy512** [comment]
I think instead, it's probably easier to look at buildHashTree() in com/google/android/snet/Filefinder.java within the SystemPartitionFileFinder() class

I'm fairly certain this is, among other things, hashing directory listings - which is why masking a file on access but not hiding it from the directory listings fails.

**Comment by hahaopsmeow** [comment]
@Entropy512 great input!

Just for reference here other methods do detect xposed: http://d3adend.org/blog/?p=589

**Comment by Entropy512** [comment]
Yeah - I'm fairly certain that's not being used yet.

I'm 90% certain that what is happening when systemless xposed is active is that SystemPartitionFileFinder sees a modified app_process binary.

What I can't figure out is why some people were saying that Nexus devices running AOSP-derivative firmware were having success with Android Pay in the past few months...  The chances of someone's build system generating an IDENTICAL app_process binary are slim to none...

**Comment by chenxiaolong** [comment]
This may be a dumb brute force solution, but instead of hooking snet, couldn't we create an `app_process` wrapper that would chroot into a mounted Nexus system image whenever the UID of the process is known to be snet? Sure, it would be a waste of storage space, but could it work at least as a temporary solution?

For example, something like:

``` c
#define SYSTEM_IMAGE_PATH   "/storage/XXXX-XXXX/system.img.ext4"
#define SYSTEM_MOUNT_POINT  "/data/local/tmp/system"

mkdir(SYSTEM_MOUNT_POINT, 0755);
mount(SYSTEM_IMAGE_PATH, SYSTEM_MOUNT_POINT, "ext4", MS_RDONLY, "");
chroot(SYSTEM_MOUNT_POINT);
...
execv("/system/bin/app_process", argv);
```

EDIT: Well, chroot wouldn't work. Creating a new mount namespace and mounting the image directly to /system would be better.

**Comment by hahaopsmeow** [comment]
@Entropy512 Could we verify your idea somehow?

We need get latest snet sources and check how they actually do see Xposed installed. (I could try to do it, but I don't know yet how to get sources)

I have Nexus 5X and Nexus 6P for testing and I am ready to test.

Just for reference: Substrate - hooking C on Android - https://koz.io/android-substrate-c-hooking/

**Comment by hahaopsmeow** [comment]
Still cannot see any snet sources on Android.
- I MitM all traffic from gactory reset Android and do not see any *.snet files.
- I checked /data/dalvik-cache with TWRP and didn't find anything with snet in it. (maybe I missed...)

Could it be that architecture of whole snet has changed?

But I see that Android loads APK from here: https://www.gstatic.com/droidguard/C4C74D1D1373D60A118D28109D7FFAECA7A892F4

There is *.so file inside the APK.

**Comment by marzika** [comment]
> https://www.gstatic.com/droidguard/C4C74D1D1373D60A118D28109D7FFAECA7A892F4

Looking at the file header, it's a ZIP derivative, as you said probably an APK or JAR, however I cannot find a way to fix the file format to be openable.

**Comment by hahaopsmeow** [comment]
try:

    $ wget -O test.apk https://www.gstatic.com/droidguard/C4C74D1D1373D60A118D28109D7FFAECA7A892F4

**Comment by hahaopsmeow** [comment]
FYI: https://github.com/pylerSM/NoDeviceCheck/issues/3#issuecomment-205074242

This doesn't exists anymore. I also checked `Google Play services_v8.7.03 (2645110-238)_apkpure.com.apk` it also doesn't have strings as shown on screenshots.

**Comment by d8ahazard** [comment]
Here's the .smali of the decompiled apk listed from the gstatic link:

http://pastebin.com/YbxBJ3Xj

**Comment by hahaopsmeow** [comment]
@d8ahazard It is just a wrapper for native code calls. All native code is in `*.so` file which is hardly obfuscated. I have absolutely no idea what it does but I bet it is some kind of smart checks...

**Comment by pylerSM** [comment]
Can you guys check this? Maybe you would have more ideas, since NDK hacking is new for me :D I just got idea but I havent been successful yet.

https://github.com/devadvance/rootcloak/issues/62

**Comment by Wetzel402** [comment]
I apologize if I am over simplifying things...

Would it be possible to intercept the JSON response, modify ctsProfileMatch to true, then pass the response on to the requesting app?  Targeting the response rather than the request?

See "Read the compatibility check response":
https://developer.android.com/training/safetynet/index.html

**Comment by Entropy512** [comment]
That's what NoDeviceCheck currently does.

Doesn't work with any app that verifies the attestation server-side...  Which seems to be all of them that use SafetyNet at this point.  (Snapchat, Android Pay)

**Comment by hahaopsmeow** [comment]
So, guys... Here is what I tried so far...

I took systemless xposed and replaces `app_process` files with untouched (original)  `app_process` files from stock ROM. Then flashed patched zip. SNET - PASS.

Then I added some zero bytes to the end of original `app_process` files and flashed it again. SNET - PASS.

So, it seems that SNET doesn't check where `app_process`:
- is mounted or not
- file creation date
- file size
- hash of file content

For me it seems it checks something else in `app_process`.

Any ideas what else could be checked in `app_process`?

**Comment by pylerSM** [comment]
just anything what can point on Xposed. even app_usage has info about Xposed.

https://github.com/rovo89/Xposed/blob/master/app_main2.cpp#L38

If you have AOSP sources, you can modify app_process randomly to see what could change SNET from success to fail.

**Comment by hahaopsmeow** [comment]
@pylerSM I do have AOSP sources... But I really don't know how to build only `app_process` files.

Can you give me a tip how to do it and I will do and write report here?

**Comment by Wetzel402** [comment]
Based on the discussion I had misinterpreted the way No Device Check functions.  

I take it no one has been able to intercept the server response to modify isValidSignature to true?  I may be over simplifying again..

**Comment by stirante** [comment]
What If we could use this https://github.com/evilsocket/arminject to hook open method in google play services and if it requests app_process point it to app_process.orig? It would work on native level, so application calls and native calls would be changed.

**Comment by pylerSM** [comment]
Does anybody have simple app which uses SafeNety checks the way Android Pay use?

**Comment by What-Zit-Tooya** [comment]
@pylerSM Netmarble games detect xposed in android. Here for details https://github.com/SudoCode755/SudoHide/issues/1

**Comment by Entropy512** [comment]
I think fundamentally there is no way to have a "simple" app for something that matches how we believe Android Pay and Snapchat - since you need SOMETHING doing a server-side attestation check, which means an app that uses some sort of service you control.

**Comment by Wetzel402** [comment]
I have scottyab's app on my phone running CM13 + Xpossed with Root Cloak and No Device Check.  It is doing a device side validation check and fails.  Based on the conversation he is looking to add server side validation and needs help.

https://github.com/scottyab/safetynethelper

**Comment by NHellFire** [comment]
There's also this app that has a server check:
https://play.google.com/store/apps/details?id=com.cigital.safetynetplayground

**Comment by azaleski9** [comment]
Hello! I worked on skipping phone verification when creating Google account. It's probably the same problem as with Android Pay. B Android sends check-in request regularly using com.google.android.gsf.checkin.CheckinTask class. Maybe it's possible to spoof this request with data from device without xposed. There is sample check-in request: http://pastebin.com/SLdggG1r

**Comment by dgw** [comment]
@FastKatt317 There's a "Subscribe" button in the GitHub sidebar for doing that, no need to comment.

**Comment by FastKatt317** [comment]
Sorry, @dgw , didn't see it. Viewing on a Nexus7

**Comment by dgw** [comment]
It's at the bottom of the page on mobile, @FastKatt317, below the comment field.

![2016-09-03 23 28 27](https://cloud.githubusercontent.com/assets/164140/18228138/4164be8e-720d-11e6-89d1-37b11d47d053.png)

**Comment by nayanbhana** [comment]
Use Magisk with Systemless Xposed.

**Comment by mirh** [comment]
Couldn't this just be focused on hiding the very xposed presence first of all?
(which is, I mean, the thing you definitively need to use a xposed module, contrarily to root)

Also, these are just my 2c, but this should be more focused on ≤4.4 android versions - since everything after already has perfectly working other forms of safetynet bypass.

**Comment by ghost** [comment]
As far as I know, there is no way to pass SafetyNet with Xposed on any version. Or do you just mean in general bypassing SafetyNet, no Xposed?

**Comment by mirh** [comment]
I mean that this module should just focus on having Xposed to work with safetynet on <kitkat. 

First, because for all practical purposes, there are already tools (well, one, magisk) that take care of that for the remainder droid versions, and it'd just make for effort duplication.
 
Second, and more importantly, because regardless I don't think there's any way to bypass dm-verity, selinux and whatever without kernel support. 
Therefore, something xposed-only like this module at the very best could just handle those systems that are not required in the first place to implement those security measures.

**Comment by ghost** [comment]
Xposed does not work with SafetyNet on any version, as I said. Magisk works, but only without xposed installed, or with the module disabled (I guess that is simpler than uninstalling and reinstalling without Magisk).
As for the other point, that definitely makes sense. I hadn't realized that Lollipop changed those restrictions.
